### PR TITLE
chore: Containerize complete build & test toolchain

### PR DESCRIPTION
**Pull request by jannfis**
This PR completely revamps the build process of ArgoCD by introducing containerized build and test targets. The complete toolchain, including Go SDK and all related tools, is now stuffed into a Docker image named `argoproj/argocd-test-tools`, there is no need for a local toolchain anymore.

* For each of the containerised targets, a local (non-dockerized) target exists as well. While `make start-e2e` starts and configures the ArgoCD end-to-end testing environment within a Docker container, `make start-e2e-local` starts it locally.
* Also, I adapted the CircleCI pipelines to use the new containerized environment. This results in better performance and stability, and also less complexity.
* I also reworked the contribution guide

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -124,64 +155,55 @@ jobs:
...
+            fi
+            test -d $HOME/.kube || mkdir -p $HOME/.kube
+            kubectl config view --raw | sed -e "s/127.0.0.1:6443/${ipaddr}:6443/g" -e "s/localhost:6443/${ipaddr}:6443/g" > $HOME/.kube/config
+          environment:
+            IFACE: ens4
       - run:
-          name: Start API server
-          command: go run ./cmd/argocd-server/main.go --loglevel debug --redis localhost:6379 --insecure --dex-server http://localhost:5556 --repo-server localhost:8081 --staticassets ../argo-cd-ui/dist/app
+          name: Start E2E test server
+          command: make start-e2e
```
Comment (alexmt) [inline .circleci/config.yml]: 👍

**Review by alexmt — APPROVED** [review]
Awesome work! Happy to merge it after tools image is pushed to argoproj dockerhub

**Comment by alexec** [comment]
I have questions from my own (not great) experience of containerizing:

1. Can you still run them from IntelliJ?
1. How do you debug tests in containers when they fail?
1. Do developers need two tool chains?
1. Are they faster or slower than the native test?

**Comment by jannfis** [comment]
> I have questions from my own (not great) experience of containerizing:
> 
>     1. Can you still run them from IntelliJ?


I don't know much about IntelliJ, but you can still run everything locally, as you were used to it before. All you have to do is to use new targets in the Makefile (i.e. `make test-local` instead of `make test`). This change is not a replacement, but an addition. What it will help with is testing and building in an environment that a) resembles what is in the final ArgoCD images and b) is the same as used in our CI pipelines.

>     2. How do you debug tests in containers when they fail?

You can start a debug shell session into the container by running `make debug-test-client` (client container) or `make debug-test-server`, so the container will be started with exact the same parameters as used for running the tests.

>     3. Do developers need two tool chains?

No. If you want to use the containerized toolchain in addition, all you have to do is pull (or build) the Docker image, then you're done. I believe this will greatly help people for their first time contributions, because the outcome also for the CI pipeline will be more predictable. Also, since we are more and more external tools in various versions, getting this all set up in your local environment can become a major PITA.
 
>     4. Are they faster or slower than the native test?

There's a small performance penalty for running the tests locally.

For containerized tests on my machine:
```
real    2m18.281s
user    0m0.049s
sys     0m0.072s
```
For locally executed tests on my machine:
```
real    2m15.947s
user    2m12.759s
sys     0m29.062s
```
Running in the CI pipeline is quite much faster, tho. With the containerized tests, they take short to 3 minutes, while the traditional way we ran them took 8 to 9 minutes.

File: docs/developer-guide/contributing.md [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+For development, you can either use the fully virtualized toolchain provided as Docker images, or you can set up the toolchain on your local development machine. Due to the dynamic nature of requirements, you might want to stay with the virtualized environment.
+
+### Install required dependencies and build-tools
+
+For installing the tools required to build and test ArgoCD on your local system, we provide convinient installer scripts. By default, they will install binaries to `/usr/local/bin` on your system, which might require `root` privileges.
+
+You can change the target location by setting the `BIN` environment before running the installer scripts. For example, you can install the binaries into `~/go/bin` (which should then be the first component in your `PATH` environment, i.e. `export PATH=~/go/bin:$PATH`):
+
+```shell
+make BIN=~/go/bin install-tools-local
```
Comment (alexmt) [inline docs/developer-guide/contributing.md]: Last minor request. Can you please mention this is only for Lunux users  ?

Comment (jannfis) [reply docs/developer-guide/contributing.md]: Done, thank you!

**Pull request merged**